tree/ntuple/v7/src/RField.cxx

```diff
@@ -2287,6 +2294,195 @@ void ROOT::Experimental::RVariantField::CommitCluster()
+{
+   auto nbytesItem = fSubFields[0]->Append(value);
+   if (IsDense()) {
+      bool mask = true;
+      Detail::RColumnElement<bool> maskElement(&mask);
+      fPrincipalColumn->Append(maskElement);
+      return 1 + nbytesItem;
+   } else {
+      fNWritten++;
+      Detail::RColumnElement<ClusterSize_t> offsetElement(&fNWritten);
+      fPrincipalColumn->Append(offsetElement);
+      return sizeof(ClusterSize_t) + nbytesItem;
+   }
+}
+
+ROOT::Experimental::RClusterIndex ROOT::Experimental::RNullableField::GetItemIndex(NTupleSize_t globalIndex)
+{
+   RClusterIndex nullIndex;
+   if (IsDense()) {
+      const bool isValidItem = *fPrincipalColumn->Map<bool>(globalIndex);
+      return isValidItem ? fPrincipalColumn->GetClusterIndex(globalIndex) : nullIndex;
+   } else {
+      RClusterIndex collectionStart;
+      ClusterSize_t collectionSize;
+      fPrincipalColumn->GetCollectionInfo(globalIndex, &collectionStart, &collectionSize);
+      return (collectionSize == 0) ? nullIndex : collectionStart;
+   }
+}
+
+void ROOT::Experimental::RNullableField::AcceptVisitor(Detail::RFieldVisitor &visitor) const
+{
+   visitor.VisitNullableField(*this);
+}
+
+//------------------------------------------------------------------------------
+
+ROOT::Experimental::RUniquePtrField::RUniquePtrField(std::string_view fieldName,
+                                                     std::unique_ptr<Detail::RFieldBase> &&itemField)
+   : RNullableField(fieldName, "std::unique_ptr<" + itemField->GetType() + ">", std::move(itemField))
+{
+}
+
+std::unique_ptr<ROOT::Experimental::Detail::RFieldBase>
+ROOT::Experimental::RUniquePtrField::CloneImpl(std::string_view newName) const
+{
+   auto newItemField = fSubFields[0]->Clone(fSubFields[0]->GetName());
+   return std::make_unique<RUniquePtrField>(newName, std::move(newItemField));
+}
+
+std::size_t ROOT::Experimental::RUniquePtrField::AppendImpl(const Detail::RFieldValue &value)
+{
+   auto typedValue = value.Get<std::unique_ptr<char>>();
+   if (*typedValue) {
+      auto itemValue = fSubFields[0]->CaptureValue(typedValue->get());
+      return AppendValue(itemValue);
+   } else {
+      return AppendNull();
+   }
+}
+
+void ROOT::Experimental::RUniquePtrField::ReadGlobalImpl(NTupleSize_t globalIndex, Detail::RFieldValue *value)
+{
+   auto ptr = value->Get<std::unique_ptr<char>>();
+   bool isValidValue = !!(*ptr);
+
+   auto itemIndex = GetItemIndex(globalIndex);
+   bool isValidItem = itemIndex.GetIndex() != kInvalidClusterIndex;
+
+   Detail::RFieldValue itemValue;
+   if (isValidValue)
+      itemValue = fSubFields[0]->CaptureValue(ptr->get());
+
+   if (isValidValue && !isValidItem) {
+      ptr->release();
+      fSubFields[0]->DestroyValue(itemValue, false /* dtorOnly */);
+      return;
+   }
+
+   if (!isValidValue && isValidItem) {
+      itemValue = fSubFields[0]->GenerateValue();
+      ptr->reset(itemValue.Get<char>());
+   }
+
+   if (isValidItem)
+      fSubFields[0]->Read(itemIndex, &itemValue);
```

Comment:
I have the feeling that this can be simplified a bit (at least, for the two last `if` statements).

Reply:
I'm not quite sure if there is a substantially better way. Sure, we can nest the two last `if` statements, but I'm not sure that it makes the code simpler / more readable. I'm open to suggestions.

Reply:
See suggestion below.  Maybe we could also nest the `if` statement in lines 2436-2440 into the `if` on line 2433, but that's more optional (up to you :slightly_smiling_face:).